azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -319,27 +321,79 @@ public void updateProjectDescription(final Project project, final String descrip
     return this.projectLoader.getProjectEvents(project, results, skip);
   }
 
-  public Props getProperties(final Project project, final String source)
-      throws ProjectManagerException {
-    return this.projectLoader.fetchProjectProperty(project, source);
+  public Props getPropertiesFromFlowFile(final Flow flow, final String jobName, final String
+      flowFileName, final int flowVersion) throws ProjectManagerException {
+    File tempDir = null;
+    Props props = null;
+    try {
+      tempDir = Files.createTempDir();
+      final File flowFile = this.projectLoader.getUploadedFlowFile(flow.getProjectId(), flow
+          .getVersion(), flowFileName, flowVersion, tempDir);
+      final String path =
+          jobName == null ? flow.getId() : flow.getId() + Constants.PATH_DELIMITER + jobName;
+      props = FlowLoaderUtils.getPropsFromYamlFile(path, flowFile);
+    } catch (final Exception e) {
+      this.logger.error("Failed to get props from flow file. " + e);
+    } finally {
+      FlowLoaderUtils.cleanUpTempDir(tempDir);
+    }
+    return props;
   }
 
-  public Props getJobOverrideProperty(final Project project, final String jobName)
-      throws ProjectManagerException {
-    return this.projectLoader.fetchProjectProperty(project, jobName + ".jor");
+  public Props getProperties(final Project project, final Flow flow, final String jobName,
+      final String source) throws ProjectManagerException {
+    if (FlowLoaderUtils.isAzkabanFlowVersion20(flow.getAzkabanFlowVersion())) {
+      // Return the properties from the original uploaded flow file.
+      return getPropertiesFromFlowFile(flow, jobName, source, 1);
+    } else {
+      return this.projectLoader.fetchProjectProperty(project, source);
+    }
   }
 
-  public void setJobOverrideProperty(final Project project, final Props prop, final String jobName,
-      final User modifier)
-      throws ProjectManagerException {
-    prop.setSource(jobName + ".jor");
-    final Props oldProps =
-        this.projectLoader.fetchProjectProperty(project, prop.getSource());
+  public Props getJobOverrideProperty(final Project project, final Flow flow, final String jobName,
+      final String source) throws ProjectManagerException {
+    if (FlowLoaderUtils.isAzkabanFlowVersion20(flow.getAzkabanFlowVersion())) {
+      final int flowVersion = this.projectLoader
+          .getLatestFlowVersion(flow.getProjectId(), flow.getVersion(), source);
+      return getPropertiesFromFlowFile(flow, jobName, source, flowVersion);
+    } else {
+      return this.projectLoader
+          .fetchProjectProperty(project, jobName + Constants.JOB_OVERRIDE_SUFFIX);
+    }
+  }
 
-    if (oldProps == null) {
-      this.projectLoader.uploadProjectProperty(project, prop);
+  public void setJobOverrideProperty(final Project project, final Flow flow, final Props prop,
```

Comment:
when are the overrode properties becoming effective? Next execution of the flow or immediately?

Reply:
It depends on whether the particular job has already started in the current flow execution. If the override properties are set before the job starts to execute, then it will take effect immediately. If set after the job has started, then it will become effective in the next execution.